### DISCUSSION TITLE
Integrating Winbond W25Q128JV SPI Flash with Crazyflie

### DISCUSSION THREAD

**Discussion by amitx64** [bitcraze/discussions#1329] [Q&A]
Hello everyone,

I'm currently working on a project involving the `integration of a Winbond W25Q128JV SPI flash` with the `Crazyflie` platform. The goal is to create a `custom deck` for storing `long trajectories` of formations and executing them without interference. However, I'm facing some confusion regarding the use of the SPI1 line, especially considering simultaneous usage with the Loco deck.

Here are my main concerns and questions:

1. **SPI1 Line Usage:** Is it acceptable to utilize the SPI1 line for connecting the Winbond SPI flash while simultaneously using the Loco deck? I want to ensure that there won't be any conflicts or performance issues.
   
2. **Trajectory Execution:** The primary aim of integrating the SPI flash is to store long trajectories of formations and execute them seamlessly. How can I ensure that the read and write operations to the flash memory do not interfere with the real-time execution of stored trajectories?
   
3. **UWB Deck Interference:** Additionally, I would like to know the rate of fetching position data from the UWB deck and the position calculation loop rate. Understanding these rates will help me assess any potential interference with SPI flash operations.

I would greatly appreciate any insights, experiences, or advice from those who have worked with similar setups or have expertise in Crazyflie development. Feel free to share your thoughts on the technical aspects, best practices, or potential challenges I might encounter.

Thank you in advance for your contributions!

**Comment by tobbeanton** [comment]
> SPI1 Line Usage: Is it acceptable to utilize the SPI1 line for connecting the Winbond SPI flash while simultaneously using the Loco deck? I want to ensure that there won't be any conflicts or performance issues.

The buss will be shared and hard to say what the effect would be without trying it for real. I don't think it will be a problem though.

> Trajectory Execution: The primary aim of integrating the SPI flash is to store long trajectories of formations and execute them seamlessly. How can I ensure that the read and write operations to the flash memory do not interfere with the real-time execution of stored trajectories?

I think buffering/caching them in RAM is a good idea.

> UWB Deck Interference: Additionally, I would like to know the rate of fetching position data from the UWB deck and the position calculation loop rate. Understanding these rates will help me assess any potential interference with SPI flash operations.

The UWB rates are pretty dynamic and varies between modes so it is hard to answer. You could check the [loco.spiWr and loco.spiRe](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/logs/#loco) log variables. However testing for real is probably the best method.

**Reply by amitx64** [reply]
@tobbeanton RAM is good but limited for long formation

**Reply by amitx64** [reply]
@tobbeanton When attempting to communicate with the SPI flash memory using SPI3, the drone gets stuck and continuously prints the following line:

`SYS---------------------------------------SYS------------------------`

Here is the relevant code snippet:

```C
static bool isInit;

static void spiFlashInit()
{
  if (isInit)
  {
    return;
  }

  spi3Begin();
  // Initialize pins
  pinMode(W25QXX_CS_PIN, OUTPUT);

  DEBUG_PRINT("Winbond W25Qxx: Using SPI Interface.\n");

  if(w25qxx_init(&w25qxx, W25QXX_CS_PIN) != W25QXX_Ok)
  {
    isInit = false;
    return;
  }

  isInit = true;
}

static bool spiFlashTest()
{
  bool status = true;

  if(!isInit)
  {
    return false;
  }

  DEBUG_PRINT("SPI Flash self-test [OK]\n");

  return status;
}

static const DeckDriver spiFlashDriver = {
  .name = "spiFlash",
  .init = spiFlashInit,
  .test = spiFlashTest,
};

DECK_DRIVER(spiFlashDriver);
```

This code initializes SPI communication with the SPI flash memory using SPI3. However, during testing, the drone gets stuck, and the mentioned line is continuously printed. It seems to indicate an issue with the SPI communication.

**Reply by tobbeanton** [reply]
spi3 is normally not used but is an optional SPI interface that can be used on the deck port. Perhaps check the [usddeck.c](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/deck/drivers/src/usddeck.c) implementation for some pointers.

**Reply by amitx64** [reply]
@tobbeanton, can I use the 'spi3' line? If so, is it okay to utilize the 'deck_spi3' methods for initializing the 'spi3' line and performing SPI transactions?

**Reply by tobbeanton** [reply]
Yes

**Reply by amitx64** [reply]
@tobbeanton I found that whenever I tried to perform a transaction using the spi3 function, the firmware gets stuck at this point during booting:
```
SYS---------------------------------------SYS------------------------
```

**Reply by amitx64** [reply]
@tobbeanton Here is my basic code for reading the Winbond W25Q128 SPI flash device ID using the SPI3 line. I encountered an error while attempting to execute this code. Below is the image of the error:
```C
#define DEBUG_MODULE "FlashDeck"
#include "debug.h"
#include "flashdeck.h"
#include "deck.h"

#define W25QXX_CS_PIN DECK_GPIO_RX2

static uint16_t w25qxx_speed;

// Handler for all w25qxx operations!
W25QXX_HandleTypeDef w25qxx;

static void spiFlashTask(void* prm);
static uint8_t xchgSpi3(uint8_t dat);

/**
 * @brief  Enables CS (driving it low) of the W25Qxx
 *
 * @param  W25Qxx handle
 * @retval None
 */
static inline void cs_on(W25QXX_HandleTypeDef *w25qxx) {
  spi3BeginTransaction(w25qxx_speed);
  digitalWrite(w25qxx->cs_pin, LOW);
}

/**
 * @brief  Disables CS (driving it high) of the W25Qxx
 *
 * @param  W25Qxx handle
 * @retval None
 */
static inline void cs_off(W25QXX_HandleTypeDef *w25qxx) {
  digitalWrite(w25qxx->cs_pin, HIGH);
  spi3EndTransaction();
}

/* Exchange a byte */
static uint8_t xchgSpi3(uint8_t dat) {
    uint8_t receive;
    // Send and receive data over SPI3
    spi3Exchange(1, &dat, &receive);

    return receive;
}

static void spiFlashTask(void* prm)
{
  // TickType_t currentTime = 0, lastWakeTime = xTaskGetTickCount();
  DEBUG_PRINT("wait for SPI FLASH\n");
  systemWaitStart();
   uint8_t buf[3];
  uint8_t deviceID[3];
  buf[0] = 0x9F;

  cs_on(&w25qxx);

  xchgSpi3(buf[0]);
  // Receive device ID
  for (int i = 0; i < 3; i++)
  {
    deviceID[i] = xchgSpi3(0xA5); // Send dummy byte to receive response
  }
  cs_off(&w25qxx);

  DEBUG_PRINT("DEVICE ID: %d , %d , %d\n", deviceID[0],deviceID[1],deviceID[2]);
  while (1)
  {
    vTaskDelay(pdMS_TO_TICKS(1000));
  }
  
}


//////////////////////////////////////////////////////////////
static bool isInit;

static void spiFlashInit()
{
  if (isInit)
  {
    return;
  }

  spi3Begin();
  w25qxx_speed = SPI3_BAUDRATE_3MHZ;
  // Init pins
  pinMode(W25QXX_CS_PIN, OUTPUT);
  digitalWrite(W25QXX_CS_PIN, HIGH);
  w25qxx.cs_pin = W25QXX_CS_PIN;

   /* create spi-flash task */
    xTaskCreate(spiFlashTask, SPI_FLASH_TASK_NAME, SPI_FLASH_TASK_STACKSIZE, NULL, SPI_FLASH_TASK_PRI, NULL);

  DEBUG_PRINT("Winbond W25Qxx: Using SPI Interface.\n");

  isInit = true;
}

static bool spiFlashTest()
{
  bool status = true;

  if(!isInit)
  {
    return false;
  }

  DEBUG_PRINT("SPI Flash self-test [OK]\n");

  return status;
}

static const DeckDriver spiFlashDriver = {
  .vid = 0,
  .pid = 0,
  .name = "spiFlash",
  .usedGpio = DECK_USING_PA3,
  .usedPeriph = DECK_USING_SPI3,      // DECK_USING_SPI3    (DECK_USING_PC10 | DECK_USING_PC11| DECK_USING_IO_4)
  .init = spiFlashInit,
  .test = spiFlashTest,
};

DECK_DRIVER(spiFlashDriver);
```

![spi_flash_error](https://github.com/bitcraze/discussions/assets/79006087/f6701106-d512-45a1-a806-88489e936a73)

I tried to trace SPI transactions on the logic analyzer, but I did not receive any clock signal through SPI3:
![spi_waveform](https://github.com/bitcraze/discussions/assets/79006087/f3292739-dbce-467b-a96a-08521fdb6a99)

**Reply by tobbeanton** [reply]
Sorry, don't have any time to help. I would recommend to get the sd-card working with spi3 as a test to see that the base functionality works.

**Reply by amitx64** [reply]
@tobbeanton ok
